### ISSUE TITLE
Support synchronizing parts of world state, since Rust isn't deterministic. 

### ISSUE THREAD

**Issue by seivan** [gschup/bevy_ggrs#110]
**Is your feature request related to a problem? Please describe.**
Rust isn't deterministic across platforms, and neither are most of rust game dev libs. 

**Describe the solution you'd like**
Make ggrs support world state syncing for select component with its rollbacks from host. 

**Describe alternatives you've considered**
Using straight up UDP for syncing host and clients. 
But it's a shame, because ggrs has the right idea of using inputs + prediction. 
The only thing is missing is an out of the box solution for syncing states since there is no determinism guarantee. 

**Additional context**
Something that supports the ideas here https://developer.valvesoftware.com/wiki/Latency_Compensating_Methods_in_Client/Server_In-game_Protocol_Design_and_Optimization

**Comment by Shelim** [comment]
Any progress on this? Or maybe some insight on how to implement it in client code? :)

**Comment by gschup** [comment]
Thanks for asking! Always happy to receive ideas and comments! Please let me elaborate why I am not a fan of this idea.

While I understand why this is a popular request, this goes against the main idea of what GGPO and therefore GGRS aims to achieve. 

Implementing the message would not be extremely complex. It would require adding a new message type to GGRS that is then pushed to the client as an Event. Then the user has to adjust the state given the information in that event. 

What is tricky here are the choices made along the way. 

- How do you ensure state corrections arrive at each client in time? What if a correction packet gets dropped?

- When applying state corrections, why do all the work of rolling back and resimulating if predicted wrong? Immediately, this leads to a different type of networking.

- In a hypothetical scenario with three clients, client A received two state corrections from B and C. Which one do you apply and which do you ignore? How to insure that all clients make the same choice? 

- Lastly, enabling state corrections also opens the door for cheating (a luxury problem).

My idea here is that GGRS is not a general purpose networking solution. If your game is not deterministic, you should probably use a different networking crate. Keeping this strong assumption leaves the crate lean-"ish" with a clear use case. 

What do you think about this? Let me know! :)

**Comment by Shelim** [comment]
I am actually investigating network crates for upcoming Bevy Jam - had my share of network game code, but outside of Rust stack. The problem I see is with float's, there can be always slight mis-calculations and drifting and those values can add-up quickly and explode the system. I wonder what kind of prevention / handle mechanism of such situations can GGRS provide. :)

Right now my idea is to designate one player (first one?) the owner of the game, and the rest players will send him periodically checksums of partial game state. In case of detected desync, the owner will send correction every few frames (these should be interpolated on clients from current state when received) until the matching checksum is send back from that player. If the owner disconnects, the next player will assume its role.

My idea is too send correction as some kind of "special Input", so they will participate in rollbacking and advancing.

**Comment by seivan** [comment]
I mean is it even possible to be deterministic with rust? And if so is it cross platform? Are any of the physics engines like Rapier or the new one (forgot the name) guaranteeing determinism of any shape? It all feels very hand wavy at this point.

**Comment by gschup** [comment]
No, neither floats nor libraries like bevy or rapier guarantee determinism for all their features. A common technique used in the fighting game development scene is to use fixed-point math instead of float computations. Using GGPO or any other rollback library usually requires either writing a lot of functionality yourself to meet the determinism-requirements or wrappers and workarounds to use other libraries. 

This is the biggest hurdle in using rollback libraries like this.

**Comment by Shelim** [comment]
Actually Rapid has article regarding achieving determinism:
https://rapier.rs/docs/user_guides/rust/determinism/

They also claim that:
> The target platforms must strictly comply to the IEEE 754-2008 floating-points standard. This ensures that floating-point computations behave the same on all platforms. This include most modern mainstream processors as well as WASM targets.

**Comment by johanhelsing** [comment]
When I've been trying to use ggrs with a regular physics engine, the problem hasn't been floats, but reliance on added/changed queries and cached state that doesn't handle rolling back, and nondeterministic iteration order in queries. 

I'm rolling my own physics for that reason, but i don't avoid floats.

**Comment by ambiso** [comment]
Is there any way of joining a running game without sending world state?

**Comment by ambiso** [comment]
Some ideas:

Maybe there could be a `.add_rollback_with_resync::<Transform>()` that requires roughly `Clone + Hash + Serialize + Deserialize` and automatically detects and corrects desyncs? This could then also be used to join running games.

> How do you ensure state corrections arrive at each client in time? 

A client might be able to decide if a state correction is viable based on its ConfirmedFrameCount - if the state update is too old it can be discarded.

> What if a correction packet gets dropped?

The other peers can periodically re-send these.

> enabling state corrections also opens the door for cheating

There could maybe be an authoritative spectator server?

> In a hypothetical scenario with three clients, client A received two state corrections from B and C. Which one do you apply and which do you ignore? How to insure that all clients make the same choice?

Alternatively, where cheating is not a concern, you could use the peerId as a tie-breaker: give the largest peerId authority.

**Comment by gschup** [comment]
> Is there any way of joining a running game without sending world state?

You could think about sending a full history of inputs from all players to catch up joining clients. similar to a match replay functionality.

**Comment by ambiso** [comment]
That's fair, though this might not be viable for all games.

Maybe some functionality (e.g. the number of confirmed frames) could be exposed to allow user code to implement re-syncing?

**Comment by seivan** [comment]
Maybe I’ve misunderstood but what’s the point of this library if achieving determinism is almost impossible or a pipe dream? Why keep the lib lean if the lib doesn’t solve the problem it’s supposed to solve?

**Comment by seivan** [comment]
> I'm rolling my own physics for that reason, but i don't avoid floats.
> 

There is no determinism across platform even with Rust floats. 

> If you're using vec3 etc, make sure you enable the libm feature of glam to disable simd instructions, though.

Does that even help?

**Comment by gschup** [comment]
> Maybe I’ve misunderstood but what’s the point of this library if achieving determinism is almost impossible or a pipe dream?

I wouldn't say it's almost impossible, it is just a very specific set of assumptions that are made to give great results. It's not just the library that is lean, it is the network traffic. Lots of modern, high-profile games can be made deterministic. Take a look at Street Fighter 6 or Guilty Gear Strive.

To answer your question literally: The point of this library was for me to learn rust. I picked a cool project (GGPO) and ported it to rust as my way of learning the language.

**Comment by johanhelsing** [comment]
> > I'm rolling my own physics for that reason, but i don't avoid floats.
> 
> There is no determinism across platform even with Rust floats.
> 
> > If you're using vec3 etc, make sure you enable the libm feature of glam to disable simd instructions, though.
> 
> Does that even help?

Yes. You could probably also do it with rapier. https://rapier.rs/docs/user_guides/rust/determinism/

It's hard, yes, but doable
